### PR TITLE
RUM-5527 Add integration tests for internal sdk core

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+    @StringForgery(type = StringForgeryType.ALPHABETICAL)
+    lateinit var fakeUserId: String
+
+    @StringForgery(regex = "[A-Z][a-z]+ [A-Z]\\. [A-Z][a-z]+")
+    lateinit var fakeUserName: String
+
+    @StringForgery(regex = "[a-z]+\\.[a-z]+@[a-z]+\\.[a-z]{3}")
+    lateinit var fakeUserEmail: String
+    private var fakeUserAdditionalProperties: Map<String, Any?> = emptyMap()
+    private lateinit var stubFeature: StorageBackedFeature
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:79]: in theory it can be just `Feature`, because there is note above that this test is not related to writing data

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+
+    @StringForgery(type = StringForgeryType.ALPHABETICAL)
+    lateinit var fakeFeatureName: String
+
+    private lateinit var fakeTrackingConsent: TrackingConsent
+
+    private var featureSdkCore: FeatureSdkCore? = null
+    private var testedInternalSdkCore: InternalSdkCore? = null
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:87]: why do we need both? by looking on the `featureSdkCore = testedInternalSdkCore` below, it seems we can just have ` testedInternalSdkCore: InternalSdkCore?` and that is it, isn't it?

also it is better to use `lateinit` with `non-null` value type `InternalSdkCore`, this will remove the need of using `?` in many places.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+    @Test
+    fun must_returnCorrectDatadogContext_when_getDatadogContext() {
+        // Given
+        val expectedServiceName = fakeConfiguration.service() ?: PACKAGE_NAME
+
+        // When
+        val context = testedInternalSdkCore?.getDatadogContext()
+
+        // Then
+        assertThat(context?.sdkVersion).isEqualTo(BuildConfig.SDK_VERSION_NAME)
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:129]: having `checkNotNull(context)` will also unlock smart cast for the code below, removing the need to add `.?`

Same for other similar tests

```suggestion
        // Then
        checkNotNull(context)
        assertThat(context.sdkVersion).isEqualTo(BuildConfig.SDK_VERSION_NAME)
```

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+            }
+        }.start()
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues2.forEach { (key, value) ->
+                    it[key] = value
+                }
+            }
+        }.start()
+        testedInternalSdkCore?.setContextUpdateReceiver(fakeFeatureName) { _, _ -> countDownLatch.countDown() }
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:208]: this should go before threads are started. There is a chance that some thread may finish its execution before this receiver is set.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+        }.start()
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues2.forEach { (key, value) ->
+                    it[key] = value
+                }
+            }
+        }.start()
+        // Give some time to the first 2 threads to start
+        Thread.sleep(100)
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:253]: minor: to be more deterministic it is better to use another latch, but I guess 100ms should be enough.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                fakeKeyValues2.forEach { (key, _) ->
+                    it[key] = fakeModifiedValues[key]
+                }
+            }
+        }.start()
+        testedInternalSdkCore?.setContextUpdateReceiver(fakeFeatureName) { _, _ -> countDownLatch.countDown() }
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:261]: same: should be set before threads are started

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+        // give some time to the firs 2 threads to start
+        Thread.sleep(100)
+        Thread {
+            featureSdkCore?.updateFeatureContext(fakeFeatureName) {
+                droppedKeyValues.forEach { (key, _) ->
+                    it.remove(key)
+                }
+            }
+        }.start()
+        testedInternalSdkCore?.setContextUpdateReceiver(fakeFeatureName) { _, _ -> countDownLatch.countDown() }
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:314]: same as above

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+    @Test
+    fun mustReturnTheCorrectNetworkInfo_when_getNetworkInfo() {
+        // Given
+        val expectedConnectivity = NetworkInfo.Connectivity.NETWORK_WIFI
+
+        // When
+        val networkInfo = testedInternalSdkCore?.networkInfo
+
+        // Then
+        assertThat(networkInfo?.connectivity).isEqualTo(expectedConnectivity)
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:339]:
```suggestion
        // Then
        checkNotNull(networkInfo)
        assertThat(networkInfo.connectivity).isEqualTo(expectedConnectivity)
```

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+            fakeFeatureName,
+            getMockServerWrapper().getServerUrl()
+        )
+        fakeTrackingConsent = forge.aValueFrom(TrackingConsent::class.java)
+        fakeUserAdditionalProperties = forge.exhaustiveAttributes(excludedKeys = setOf("id", "name", "email"))
+        testedInternalSdkCore = Datadog.initialize(
+            ApplicationProvider.getApplicationContext(),
+            fakeConfiguration,
+            fakeTrackingConsent
+        ) as? InternalSdkCore
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:107]:
```suggestion
        ) as InternalSdkCore
```

We want test to fail here if this cannot be casted. It will be closer to the failure site.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:1]: are these test actually different from what we have for the `CoreFeature` in the unit-tests? I have the feeling that we are repeating the stuff with a little added value here.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/tests/utils/MutableMapExt.kt [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+package com.datadog.android.core.integration.tests.utils
+
+import fr.xgouchet.elmyr.Forge
+
+internal fun <T : Any, R : Any?> MutableMap<T, R>.removeRandomEntries(forge: Forge): Map<Any, Any?> {
+    val removedEntries = mutableMapOf<Any, Any?>()
+    val keys = this.keys.toList()
+    val numberOfElementsToRemove = forge.anInt(1, keys.size)
+    synchronized(this) {
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/tests/utils/MutableMapExt.kt:15]: was thinking just in case someone is using without making sure it is thread safe. I guess it is too much, will remove that lock.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 11]
```diff
@@ -0,0 +1,626 @@
...
+import java.io.File
+import java.util.UUID
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+
+/**
+ * Provides the tests for the InternalSdkCore API not related with the writing operations.
+ */
+@RunWith(AndroidJUnit4::class)
+class InternalSdkCoreContextTest : MockServerTest() {
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:65]: it is a convention I had with Xav, we are not sure it is the best name there TBH :). The intention is to test first the static APIs in the InternalSdkCore and SdkCore (the ones that do not interact with the writer and for which we do not need a MockWebServer) and the other tests will be added in separated classes.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt (hunk 2 above)
Comment (mariusc83) [reply reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:87]: we also need the feature to assert it below where we check the number of returned features.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt
Comment (mariusc83) [reply reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:1]: The intentions is to test all the public APIs, we do not care if the functionality was already tested somewhere else. Otherwise we will miss things and will be hard to keep track.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt (hunk 2 above)
Comment (0xnm) [reply reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:87]: I'm not sure I got it. `InternalSdkCore` extends `FeatureSdkCore`, so having just `testedInternalSdkCore: InternalSdkCore` should be sufficient to do the necessary assertions.

Comment (mariusc83) [reply reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:87]: oh yes, sorry you're totally right, not sure why I added that.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 12]
```diff
@@ -0,0 +1,625 @@
...
+
+    @StringForgery(regex = "[a-z]+\\.[a-z]+@[a-z]+\\.[a-z]{3}")
+    lateinit var fakeUserEmail: String
+    private var fakeUserAdditionalProperties: Map<String, Any?> = emptyMap()
+    private lateinit var stubFeature: Feature
+
+    @StringForgery(type = StringForgeryType.ALPHABETICAL)
+    lateinit var fakeFeatureName: String
+
+    private lateinit var fakeTrackingConsent: TrackingConsent
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:82]: minor: you can simply add `Forgery` annotation here and remove `lateinit` and `fakeTrackingConsent = forge.aValueFrom(TrackingConsent::class.java)` line.

File: reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt [hunk 13]
```diff
@@ -0,0 +1,625 @@
...
+        } else {
+            expectedRegisteredFeatures.size
+        }
+
+        // When
+        val features = testedInternalSdkCore.getAllFeatures().map { it.unwrap<Feature>() }
+
+        // Then
+        assertThat(features).containsAll(expectedRegisteredFeatures)
+        assertThat(features.size).isEqualTo(expectedRegisteredFeaturesSize)
```
Comment (0xnm) [inline reliability/core-it/src/androidTest/kotlin/com/datadog/android/core/integration/InternalSdkCoreContextTest.kt:537]:
```suggestion
        assertThat(features).hasSize(expectedRegisteredFeaturesSize)
```

**Review by 0xnm — APPROVED** [review]
